stl/src/stacktrace.cpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <shared_mutex>
+#include <string>
+
+// clang-format off
+#include <Windows.h>
+#include <dbghelp.h>
+// clang-format on
+
+#pragma comment(lib, "Dbghelp.lib")
+
+namespace {
+    static constexpr std::size_t max_line_size = 2000;
+
+    static constexpr std::size_t table_size_power_of_two_factor = 10;
+
+    static constexpr std::size_t table_size = 1 << table_size_power_of_two_factor;
+
+    static constexpr SYMBOL_INFO init_symbol_info() {
+        SYMBOL_INFO result  = {};
+        result.SizeOfStruct = sizeof(SYMBOL_INFO);
+        result.MaxNameLen   = max_line_size;
+        return result;
+    }
+
+    struct stacktrace_address_table_entry_t {
+        void* address             = nullptr;
+        bool is_description_valid = false;
+        bool is_line_valid        = false;
+        IMAGEHLP_LINE line        = {sizeof(IMAGEHLP_LINE)};
+        SYMBOL_INFO info          = init_symbol_info();
+        wchar_t buffer[max_line_size];
+    };
+
+    static stacktrace_address_table_entry_t stacktrace_address_table[table_size];

```

Comment:
Why is this global state needed?

Reply:
Generally global state here needed for `OpenProcess`+`SymInitialize` / `SymCleanup`+`CloseHandle`.

This particular global state is a cache. The decision towards making a cache was guided by these facts:
 * Obtaining information should be lazy
 * We anyway have to lock, as APIs are not thread safe
 * We have a random-access array of stacktrace entries, which are returned by ref, meaning they should be plain pointers
 * My pas experience (although in Windows XP times) shows these APIs are slow
 * There's `source_file()` and `source_line()` which are likely to be called in sequence, so it is natural to cache at least one last entry
 * Making cache to be more than just one entry is almost free

Reply:
I don't think it's useful to cache this information. I think that the main use for `stacktrace` is to help debugging. It is likely to be used only in case of exception. `stacktrace` doesn't have to be fast, it should be optimized for the case when it's never used (since most programs are meant not to exit with an exception). This cache is huge (if my calculations are correct it has 1024 entries each of which has an array of 2000 `wchar_t`s).

Reply:
Removed the cache table, still cache last address.

Reply:
Removed caching entirely.